### ISSUE TITLE
mixed numeric and data operations (`restriction` group)

### ISSUE THREAD

**Issue by Rudxain** [rust-lang/rust-clippy#17649]
### What it does

It triggers when treating a number (float, int, complex, ...) as both raw-data and as a mathematical value, or when operating two or more numeric-values in such a way that some are treated as data and others as numbers (with exceptions, such as `x << y`, because LHS must be data and RHS must be num, so this won't fire the lint)

Note that "raw data operation" isn't just "bitwise operators", it can be (safe) type-casts and (unsafe) `transmute`

### Advantage

Type safety: it forces the user to think about which values should be actual numbers and which are "just bits". This is so that there's no accidental misuse.

### Drawbacks

In some cases, too many false-positives. Hence `restriction`

### Example

```rust
let mut a: u16 = 0xaa55;
a >>= 1;
a += 1;
```

Perhaps the user meant:

```rust
let mut a: u16 = 0xaa55;
a >>= 1;
a |= 1;
```

### Comparison with existing lints

- `*_endian_bytes` is just a niche subset of this concept, but it's not really a "mixed operation" IMO
- `precedence` is purely local, it looks at individual expressions. My proposal takes into account a bigger context: from the var declaration to every expression that references it.

### Additional Context

This is somewhat related to #16213. That was part of the inspiration, but I've always wanted Rust to have an opt-in "mode" that distinguished between opaque mathematical ints and bit-sequences.

Allowing mixed operations is very convenient for a low-level lang such as Rust, but this can be bad in some safety-critical situations.

An alternative to this lint would be for the user to define opaque newtypes that only expose the desired operations, and use macros for syntax-sugar. I'm unsure if this would be better or worse

**Comment by Jarcho** [comment]
Interacting with numbers at the bit level frequently does involve both arithmetic and bitwise operations. Accessing specific bits is about the only thing that doesn't.
